Sources/SPMBuildCore/ExtensionEvaluator.swift

```diff
@@ -0,0 +1,328 @@
+        
+        /// A command to run after the end of every build.
+        case postbuildCommand(
+                displayName: String,
+                execPath: AbsolutePath,
+                arguments: [String],
+                workingDir: AbsolutePath?,
+                environment: [String: String]?
+             )
+    }
+    
+    // Any diagnostics emitted by the extension.
+    public let diagnostics: [Diagnostic]
+    
+    // A location representing a file name or path and an optional line number.
+    // FIXME: This should be part of the Diagnostics APIs.
+    struct FileLineLocation: DiagnosticLocation {
+        let file: String
+        let line: Int?
+        var description: String {
+            "\(file)\(line.map{":\($0)"} ?? "")"
+        }
+    }
+    
+    // Any textual output emitted by the extension.
+    public let textOutput: String
+}
+
+
+/// An error in extension evaluation.
+public enum ExtensionEvaluationError: Swift.Error {
+    case outputDirectoryCouldNotBeCreated(path: AbsolutePath, underlyingError: Error)
+    case runningExtensionFailed(underlyingError: Error)
+    case decodingExtensionOutputFailed(json: Data, underlyingError: Error)
+}
+
+
+/// Implements the mechanics of running an extension script (implemented as a set of Swift source files) as a process.
+public protocol ExtensionRunner {
+    
+    /// Implements the mechanics of running an extension script implemented as a set of Swift source files, for use
+    /// by the package graph when it is evaluating package extensions.
+    ///
+    /// The `sources` refer to the Swift source files and are accessible in the provided `fileSystem`. The input is
+    /// a serialized ExtensionEvaluationContext, and the output should be a serialized ExtensionEvaluationOutput as
+    /// well as any free-form output produced by the script (for debugging purposes).
+    ///
+    /// Any errors or warnings related to the running of the extension will be added to `diagnostics`.  Any errors
+    /// or warnings emitted by the extension itself will be part of the returned output.
+    ///
+    /// Every concrete implementation should cache any intermediates as necessary for fast evaluation.
+    func runExtension(
+        sources: Sources,
+        inputJSON: Data,
+        diagnostics: DiagnosticsEngine,
+        fileSystem: FileSystem
+    ) throws -> (outputJSON: Data, stdoutText: Data)
+}
+
+
+/// Serializable context that's passed as input to the evaluation of the extension.
+struct ExtensionEvaluationInput: Codable {
+    var targetName: String
+    var moduleName: String
+    var targetDir: String
+    var packageDir: String
+    var sourceFiles: [String]
+    var dependencies: [DependencyTarget]
+    public struct DependencyTarget: Codable {
+        var targetName: String
+        var moduleName: String
+        var targetDir: String
+    }
+    var outputDir: String
+    var cacheDir: String
+    var execsDir: String
+    var options: [String: String]
```

Comment:
should this be optional?

Reply:
Yes, or probably removed since we're not doing options initially.